syntax = "proto3";

package eventtypes;

import public "gogo.proto";
import "google/protobuf/descriptor.proto";
import "attributes.proto";

extend google.protobuf.EnumValueOptions {
    // event severity
    eventattrs.Severity severity   = 10005;

    // event category
    eventattrs.Category category   = 10006;

    // description/UI hint of the event
    string desc                    = 10007;

    // indicates whether an event should be suppressed during maintenance mode
    bool suppress_mm               = 10008;
}

enum EventTypes {
    option (gogoproto.goproto_enum_stringer) = false;
    option (gogoproto.enum_stringer) = true;
    option (gogoproto.enum_customname) = "EventType";
    option (gogoproto.goproto_enum_prefix) = false;

    // ----------------------------- Cluster events -------------------------- //
    ELECTION_STARTED                = 0 [(category)    = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Leader election started in the cluster",
                                         (suppress_mm) = true];

    ELECTION_CANCELLED              = 1 [(category)    = Cluster,
                                         (severity)    = WARN,
                                         (desc)        = "Leader election cancelled",
                                         (suppress_mm) = true];

    ELECTION_NOTIFICATION_FAILED    = 2 [(category)    = Cluster,
                                         (severity)    = WARN,
                                         (desc)        = "Failed to send leader election notification",
                                         (suppress_mm) = true];

    ELECTION_STOPPED                = 3 [(category)    = Cluster,
                                         (severity)    = WARN,
                                         (desc)        = "Leader election stopped",
                                         (suppress_mm) = true];

    LEADER_ELECTED                  = 4 [(category)    = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Leader elected for the cluster",
                                         (suppress_mm) = true];

    LEADER_LOST                     = 5 [(category)    = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Node lost leadership during the election",
                                         (suppress_mm) = true];

    LEADER_CHANGED                  = 6 [(category)    = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Leader changed in the election",
                                         (suppress_mm) = true];


    // ------------------------------- Node events ----------------------------- //
    NODE_JOINED                     = 7 [(category)    = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Node joined the cluster",
                                         (suppress_mm) = true];

    NODE_DISJOINED                  = 8 [(category)    = Cluster,
                                         (severity)    = WARN,
                                         (desc)        = "Node disjoined from the cluster"];

    NODE_HEALTHY                    = 9 [(category)    = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Node is healthy",
                                         (suppress_mm) = true];

    NODE_UNREACHABLE                = 10 [(category)    = Cluster,
                                          (severity)    = CRITICAL,
                                          (desc)        = "Node is unreachable",
                                          (suppress_mm) = true];

    // ------------------------------- Quorum events ----------------------------- //
    QUORUM_MEMBER_ADD               = 15 [(category)   = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Added member to quorum",
                                         (suppress_mm) = true];

    QUORUM_MEMBER_REMOVE            = 16 [(category)   = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Removed member from quorum",
                                         (suppress_mm) = true];

    QUORUM_MEMBER_HEALTHY           = 17 [(category)   = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Quorum member is now healthy",
                                         (suppress_mm) = true];

    QUORUM_MEMBER_UNHEALTHY         = 18 [(category)   = Cluster,
                                         (severity)    = CRITICAL,
                                         (desc)        = "Quorum member is now unhealthy",
                                         (suppress_mm) = true];

    UNSUPPORTED_QUORUM_SIZE         = 19 [(category)   = Cluster,
                                         (severity)    = CRITICAL,
                                         (desc)        = "Quorum size is below supported minimum",
                                         (suppress_mm) = true];

    QUORUM_UNHEALTHY                = 20 [(category)   = Cluster,
                                         (severity)    = CRITICAL,
                                         (desc)        = "Quorum does not have enough healthy members",
                                         (suppress_mm) = true];

    // ------------------------------- Diagnostic events ----------------------------- //
    MODULE_CREATION_FAILED          = 24 [(category)   = Cluster,
                                         (severity)    = WARN,
                                         (desc)        = "Module creation for diagnostics failed",
                                         (suppress_mm) = true];

    // ------------------------ Config Snapshot/Restore events ------------------------ //
    CONFIG_RESTORED                 = 30 [(category)   = Cluster,
                                         (severity)    = INFO,
                                         (desc)        = "Configuration was restored",
                                         (suppress_mm) = true];

    CONFIG_RESTORE_ABORTED          = 31 [(category)   = Cluster,
                                         (severity)    = WARN,
                                         (desc)        = "Configuration restore operation was aborted",
                                         (suppress_mm) = true];

    CONFIG_RESTORE_FAILED           = 32 [(category)   = Cluster,
                                         (severity)    = CRITICAL,
                                         (desc)        = "Configuration restore operation failed",
                                         (suppress_mm) = true];


    // -------------------------------- Host/DSC events -------------------------- //
    DSC_ADMITTED                    = 100 [(category)    = Cluster,
                                           (severity)    = INFO,
                                           (desc)        = "DSC admitted to the cluster",
                                           (suppress_mm) = true];

    DSC_REJECTED                    = 101 [(category)    = Cluster,
                                           (severity)    = WARN,
                                           (desc)        = "DSC admission request was rejected"];

    DSC_UNREACHABLE                 = 102 [(category)    = Cluster,
                                           (severity)    = CRITICAL,
                                           (desc)        = "DSC is unreachable",
                                           (suppress_mm) = true];

    DSC_HEALTHY                     = 103 [(category)    = Cluster,
                                           (severity)    = INFO,
                                           (desc)        = "DSC is healthy",
                                           (suppress_mm) = true];

    DSC_UNHEALTHY                   = 104 [(category)    = Cluster,
                                           (severity)    = CRITICAL,
                                           (desc)        = "DSC is unhealthy",
                                           (suppress_mm) = true];

    HOST_DSC_SPEC_CONFLICT          = 105 [(category)    = Cluster,
                                           (severity)    = WARN,
                                           (desc)        = "The system has detected a conflict between the DSC specifications of different Host objects"];

    DSC_DEADMITTED                  = 106 [(category)    = Cluster,
                                           (severity)    = INFO,
                                           (desc)        = "DSC de-admitted from the cluster"];

    DSC_DECOMMISSIONED              = 107 [(category)    = Cluster,
                                           (severity)    = INFO,
                                           (desc)        = "DSC decommissioned from the cluster"];

    // ----------------------------- API Gateway events ---------------------- //
    AUTO_GENERATED_TLS_CERT         = 200 [(category)    = Cluster,
                                           (severity)    = WARN,
                                           (desc)        = "Auto generated certificate is being used for API Gateway TLS",
                                           (suppress_mm) = true];

    // --------------------------- Auth/Audit events ------------------------- //
    LOGIN_FAILED                    = 300 [(category)    = Cluster,
                                           (severity)    = INFO,
                                           (desc)        = "User login failed"];

    AUDITING_FAILED                 = 301 [(category)    = Cluster,
                                           (severity)    = CRITICAL,
                                           (desc)        = "Writing of AuditEvent failed",
                                           (suppress_mm) = true];

    PASSWORD_CHANGED                = 302 [(category)    = Cluster,
                                           (severity)    = WARN,
                                           (desc)        = "Password changed"];

    PASSWORD_RESET                  = 303 [(category)    = Cluster,
                                           (severity)    = WARN,
                                           (desc)        = "Password reset"];


    // --------------------------- HAL/Link events --------------------------- //
    LINK_UP                         = 400 [(category)    = Network,
                                           (severity)    = INFO,
                                           (desc)        = "Port is linked up",
                                           (suppress_mm) = true];

    LINK_DOWN                       = 401 [(category)    = Network,
                                           (severity)    = WARN,
                                           (desc)        = "Port link status is down",
                                           (suppress_mm) = true];


    // ------------------------------ System events -------------------------- //
    SERVICE_STARTED                 = 500 [(category)    = System,
                                           (severity)    = DEBUG,
                                           (desc)        = "Service started",
                                           (suppress_mm) = true];

    SERVICE_STOPPED                 = 501 [(category)    = System,
                                           (severity)    = WARN,
                                           (desc)        = "Service stopped",
                                           (suppress_mm) = true];

    NAPLES_SERVICE_STOPPED          = 502 [(category)    = System,
                                           (severity)    = CRITICAL,
                                           (desc)        = "Naples service stopped",
                                           (suppress_mm) = true];

    SERVICE_PENDING                 = 503 [(category)    = System,
                                           (severity)    = DEBUG,
                                           (desc)        = "Service pending"];

    SERVICE_RUNNING                 = 504 [(category)    = System,
                                           (severity)    = DEBUG,
                                           (desc)        = "Service running",
                                           (suppress_mm) = true];

    SERVICE_UNRESPONSIVE            = 505 [(category)    = System,
                                           (severity)    = CRITICAL,
                                           (desc)        = "Service unresponsive due to lack of system resources"];

    SYSTEM_COLDBOOT                 = 506 [(category)    = System,
                                           (severity)    = WARN,
                                           (desc)        = "System cold booted",
                                           (suppress_mm) = true];

    SYSTEM_RESOURCE_USAGE           = 507 [(category)    = System,
                                           (severity)    = WARN,
                                           (desc)        = "System resource usage is high"];

    NAPLES_FATAL_INTERRUPT          = 508 [(category)    = System,
                                           (severity)    = CRITICAL,
                                           (desc)        = "Naples has a fatal interrupt"];

    NAPLES_CATTRIP_INTERRUPT        = 509 [(category)    = System,
                                           (severity)    = CRITICAL,
                                           (desc)        = "System encountered cattrip resetting system"];

    NAPLES_OVER_TEMP                = 510 [(category)    = System,
                                           (severity)    = WARN,
                                           (desc)        = "System temperature is above threshold."];

    NAPLES_OVER_TEMP_EXIT           = 511 [(category)    = System,
                                           (severity)    = WARN,
                                           (desc)        = "System temperature is below threshold."];

    NAPLES_PANIC_EVENT              = 512 [(category)    = System,
                                           (severity)    = WARN,
                                           (desc)        = "System panic on the previous boot"];

    NAPLES_POST_DIAG_FAILURE_EVENT  = 513 [(category)    = System,
                                           (severity)    = WARN,
                                           (desc)        = "System post diag test failed"];

    // ------------------------------ Rollout events -------------------------- //
    ROLLOUT_STARTED                 = 701 [(category) = Rollout,
                                           (severity) = INFO,
                                           (desc)     = "Rollout initiated"];

    ROLLOUT_SUCCESS                 = 702 [(category) = Rollout,
                                           (severity) = INFO,
                                           (desc)     = "Rollout successfully completed"];

    ROLLOUT_FAILED                  = 703 [(category) = Rollout,
                                           (severity) = CRITICAL,
                                           (desc)     = "Rollout failed"];

    ROLLOUT_SUSPENDED               = 704 [(category) = Rollout,
                                           (severity) = INFO,
                                           (desc)     = "Rollout suspended"];

    // ------------------------------ Config events -------------------------- //
    CONFIG_FAIL                     = 801 [(category) = Config,
                                           (severity) = WARN,
                                           (desc)     = "Configuration failed"];

}
